### PR TITLE
Mask Generation Task Guide

### PR DESCRIPTION
**Pull request by merveenoyan**
Added mask generation task guide.

### CODE REVIEW

File: docs/source/en/tasks/mask_generation.md [hunk 1]
```diff
@@ -0,0 +1,247 @@
+<!--Copyright 2023 The HuggingFace Team. All rights reserved.
```
Comment (NielsRogge) [inline docs/source/en/tasks/mask_generation.md:1]:
```suggestion
<!--Copyright 2024 The HuggingFace Team. All rights reserved.
```

File: docs/source/en/tasks/mask_generation.md [hunk 2]
```diff
@@ -0,0 +1,247 @@
...
+⚠️ Note that this file is in Markdown but contain specific syntax for our doc-builder (similar to MDX) that may not be
+rendered properly in your Markdown viewer.
+
+-->
+
+# Mask Generation
+
+Mask generation is the task of generating semantically meaningful masks for an image. 
+This task is very similar to image segmentation, but many differences exist. Image segmentation models are trained
+on labeled datasets and are limited to the classes they have seen during training; they return a mask and its class,
```
Comment (NielsRogge) [inline docs/source/en/tasks/mask_generation.md:21]:
```suggestion
on labeled datasets and are limited to the classes they have seen during training; they return a set of masks and corresponding classes, 
```

File: docs/source/en/tasks/mask_generation.md [hunk 3]
```diff
@@ -0,0 +1,247 @@
...
+
+Mask generation is the task of generating semantically meaningful masks for an image. 
+This task is very similar to image segmentation, but many differences exist. Image segmentation models are trained
+on labeled datasets and are limited to the classes they have seen during training; they return a set of masks and corresponding classes, 
+given an image. 
+
+Mask generation models are trained on large amounts of data, and they operate in two modes. The first one is prompting 
+mode. In this mode, the model takes in an image and a prompt, where a prompt can be a point location in the image within 
+an object or a bounding box surrounding an object. In prompting mode, the model only returns the mask over the object 
+that the prompt is pointing out to. The second one is the  segment everything mode. In segment everything, given an
```
Comment (NielsRogge) [inline docs/source/en/tasks/mask_generation.md:27]:
```suggestion
that the prompt is pointing out to. The second one is the segment everything mode. In segment everything, given an 
```

File: docs/source/en/tasks/mask_generation.md [hunk 4]
```diff
@@ -0,0 +1,247 @@
...
+
+First, let's install `transformers`:
+
+```bash
+pip install -q transformers
+```
+
+## Mask Generation Pipeline
+
+The easiest way to infer mask generation models is to use `mask-generation` pipeline.
```
Comment (NielsRogge) [inline docs/source/en/tasks/mask_generation.md:56]:
```suggestion
The easiest way to infer mask generation models is to use the `mask-generation` pipeline.
```

File: docs/source/en/tasks/mask_generation.md [hunk 5]
```diff
@@ -0,0 +1,247 @@
...
+
+⚠️ Note that this file is in Markdown but contain specific syntax for our doc-builder (similar to MDX) that may not be
+rendered properly in your Markdown viewer.
+
+-->
+
+# Mask Generation
+
+Mask generation is the task of generating semantically meaningful masks for an image. 
+This task is very similar to image segmentation, but many differences exist. Image segmentation models are trained
```
[the remark this answers is not in the document]
Comment (merveenoyan) [reply docs/source/en/tasks/mask_generation.md:20]: I'll fix this, it should've been semantic_segmentation so

**Review by MKhalusova — COMMENTED** [review]
Great work, @merveenoyan ! Left some suggestions.

File: docs/source/en/tasks/mask_generation.md [hunk 6]
```diff
@@ -0,0 +1,247 @@
...
+on labeled datasets and are limited to the classes they have seen during training; they return a set of masks and corresponding classes, 
+given an image. 
+
+Mask generation models are trained on large amounts of data, and they operate in two modes. The first one is prompting
```
[the remark this answers is not in the document]
Comment (merveenoyan) [reply docs/source/en/tasks/mask_generation.md:24]: the model is actually manually labeled with masks, semi-automatic labeled, and fully automatically labeled, however the label here doesn't imply class masks like fully supervised image segmentation. will try to put this properly

File: docs/source/en/tasks/mask_generation.md [hunk 7]
```diff
@@ -0,0 +1,247 @@
...
+
+# Mask Generation
+
+Mask generation is the task of generating semantically meaningful masks for an image. 
+This task is very similar to image segmentation, but many differences exist. Image segmentation models are trained
+on labeled datasets and are limited to the classes they have seen during training; they return a set of masks and corresponding classes, 
+given an image. 
+
+Mask generation models are trained on large amounts of data, and they operate in two modes. The first one is prompting 
+mode. In this mode, the model takes in an image and a prompt, where a prompt can be a point location in the image within
```
[the remark this answers is not in the document]
Comment (merveenoyan) [reply docs/source/en/tasks/mask_generation.md:25]: yes it is, will clarify

**Comment by merveenoyan** [comment]
@MKhalusova thanks, I addressed your comments I think

**Review by MKhalusova — APPROVED** [review]
Thanks for addressing the feedback, LGTM!

**Comment by MKhalusova** [comment]
@ArthurZucker please take a look. the guide LGTM

File: docs/source/en/tasks/mask_generation.md [hunk 8]
```diff
@@ -0,0 +1,240 @@
...
+rendered properly in your Markdown viewer.
+
+-->
+
+# Mask Generation
+
+Mask generation is the task of generating semantically meaningful masks for an image. 
+This task is very similar to [image segmentation](semantic_segmentation), but many differences exist. Image segmentation models are trained on labeled datasets and are limited to the classes they have seen during training; they return a set of masks and corresponding classes, given an image. 
+
+Mask generation models are trained on large amounts of data and operate in two modes.
```
Comment (khipp) [inline docs/source/en/tasks/mask_generation.md:22]:
```suggestion
Mask generation models are trained on large amounts of data and operate in two modes:
```

File: docs/source/en/tasks/mask_generation.md [hunk 9]
```diff
@@ -0,0 +1,240 @@
...
+Let's segment everything. `points-per-batch` enables parallel inference of points in
+segment everything mode. This enables faster inference, but consumes more memory.
+Moreover, SAM only enables batching over points and not the images. `pred_iou_thresh` is
+the IoU confidence threshold where only the masks above that certain threshold are returned.
+
+```python
+masks = mask_generator(image, points_per_batch=128, pred_iou_thresh=0.88)
+```
+
+The `masks` looks like following:
```
Comment (khipp) [inline docs/source/en/tasks/mask_generation.md:81]:
```suggestion
The `masks` looks like the following:
```

File: docs/source/en/tasks/mask_generation.md [hunk 10]
```diff
@@ -0,0 +1,240 @@
...
+
+plt.show()
+```
+<div class="flex justify-center">
+     <img src="https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/transformers/tasks/masks.png" alt="Visualized"/>
+</div>
+
+### Box Prompting
+
+You can also do box prompting in a similar fashion to point prompting. You can simply pass the input box in format of a list
```
Comment (khipp) [inline docs/source/en/tasks/mask_generation.md:183]:
```suggestion
You can also do box prompting in a similar fashion to point prompting. You can simply pass the input box in the format of a list
```

File: docs/source/en/tasks/mask_generation.md [hunk 11]
```diff
@@ -0,0 +1,240 @@
...
+`[x_min, y_min, x_max, y_max]` format along with the image to the `processor`. Take the processor output and directly pass it 
+to the model, then postprocess the output again.
```
Comment (khipp) [inline docs/source/en/tasks/mask_generation.md:185]: You used "post-process" in line 138. I would keep it consistent and stick to either.

File: docs/source/en/tasks/mask_generation.md [hunk 12]
```diff
@@ -0,0 +1,240 @@
...
+```
+
+<div class="flex justify-center">
+     <img src="https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/bee.jpg" alt="Example Image"/>
+</div>
+
+Let's segment everything. `points-per-batch` enables parallel inference of points in
+segment everything mode. This enables faster inference, but consumes more memory.
+Moreover, SAM only enables batching over points and not the images. `pred_iou_thresh` is
+the IoU confidence threshold where only the masks above that certain threshold are returned.
```
Comment (ArthurZucker) [inline docs/source/en/tasks/mask_generation.md:75]: should we just not split this and leave it to the window manager?

File: docs/source/en/tasks/mask_generation.md [hunk 13]
```diff
@@ -0,0 +1,240 @@
...
+`original_sizes` and `reshaped_input_sizes` we take from the processor's initial output. We need to pass these 
+since processor resizes the image, and the output needs to be extrapolated.
+
+```
+input_points = [[[2592, 1728]]] # point location of the bee
+
+inputs = processor(image, input_points=input_points, return_tensors="pt").to(device)
+outputs = model(**inputs)
+masks = processor.image_processor.post_process_masks(outputs.pred_masks.cpu(), inputs["original_sizes"].cpu(), inputs["reshaped_input_sizes"].cpu())
+```
```
Comment (ArthurZucker) [inline docs/source/en/tasks/mask_generation.md:148]:
```suggestion

```python
input_points = [[[2592, 1728]]] # point location of the bee

inputs = processor(image, input_points=input_points, return_tensors="pt").to(device)
with torch.no_grad():
    outputs = model(**inputs)
masks = processor.image_processor.post_process_masks(outputs.pred_masks.cpu(), inputs["original_sizes"].cpu(), inputs["reshaped_input_sizes"].cpu())
```
```

**Review by ArthurZucker — APPROVED** [review]
Thanks for the nice documentation!

File: docs/source/en/tasks/mask_generation.md (hunk 8 above)
Comment (merveenoyan) [reply docs/source/en/tasks/mask_generation.md:22]: I would rather not have this.

File: docs/source/en/tasks/mask_generation.md (hunk 12 above)
Comment (ArthurZucker) [reply docs/source/en/tasks/mask_generation.md:75]: No No i just meant that you have 4lines here
```suggestion
Let's segment everything. `points-per-batch` enables parallel inference of points in segment everything mode. This enables faster inference, but consumes more memory. Moreover, SAM only enables batching over points and not the images. `pred_iou_thresh` is the IoU confidence threshold where only the masks above that certain threshold are returned.
```

Comment (ArthurZucker) [reply docs/source/en/tasks/mask_generation.md:75]: 😈

File: docs/source/en/tasks/mask_generation.md [hunk 14]
```diff
@@ -0,0 +1,237 @@
...
+<div class="flex justify-center">
+     <img src="https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/transformers/tasks/masks.png" alt="Visualized"/>
+</div>
+
+### Box Prompting
+
+You can also do box prompting in a similar fashion to point prompting. You can simply pass the input box in the format of a list
+`[x_min, y_min, x_max, y_max]` format along with the image to the `processor`. Take the processor output and directly pass it 
+to the model, then postprocess the output again.
```
Comment (merveenoyan) [inline docs/source/en/tasks/mask_generation.md:182]:
```suggestion
to the model, then post-process the output again.
```

**Comment by merveenoyan** [comment]
@ArthurZucker can you merge this if it's ok?

File: docs/source/en/tasks/mask_generation.md [hunk 15]
```diff
@@ -0,0 +1,237 @@
...
+since the processor resizes the image, and the output needs to be extrapolated.
+
+```python
+input_points = [[[2592, 1728]]] # point location of the bee
+
+inputs = processor(image, input_points=input_points, return_tensors="pt").to(device)
+with torch.no_grad():
+    outputs = model(**inputs)
+masks = processor.image_processor.post_process_masks(outputs.pred_masks.cpu(), inputs["original_sizes"].cpu(), inputs["reshaped_input_sizes"].cpu())
+
```
Comment (merveenoyan) [inline docs/source/en/tasks/mask_generation.md:146]:
```suggestion
```
```
